### PR TITLE
provide useful error when invalid service name provided

### LINKED ISSUE #2494: firewalld Traceback error instead of something meaningful
##### ISSUE TYPE

- Bug Report
##### COMPONENT NAME

firewalld
##### ANSIBLE VERSION

```
ansible 2.1.0.0
```
##### OS / ENVIRONMENT

RHEL 7.2
##### SUMMARY

Traceback error instead of "2016-06-28 13:00:33 ERROR: INVALID_SERVICE: httpd"
##### STEPS TO REPRODUCE

try to manage an invalid port or service

```
    firewalld:
      service: httpd
      immediate: true
      permanent: true
      state: enabled
```

##### EXPECTED RESULTS

An error that is meaningful like that of firewalld service on the node:
"2016-06-28 13:00:33 ERROR: INVALID_SERVICE: httpd"
##### ACTUAL RESULTS

A Traceback error that is difficult to understand what the problem is.
...

```
[ansible01.example.org]: FAILED! => {"changed": false, "failed": true, "module_stderr": "", "module_stdout": "Traceback (most recent call last):\r\n  File \"/tmp/ansible_fClbvK/ansible_module_firewalld.py\", line 605, in <module>\r\n    main()\r\n  File \"/tmp/ansible_fClbvK/ansible_module_firewalld.py\", line 403, in main\r\n
```

### PR DESCRIPTION
**Pull request by maxamillion**
##### ISSUE TYPE

- Feature Pull Request
##### COMPONENT NAME

firewalld
##### ANSIBLE VERSION

```
ansible 2.3.0 (devel 7437bd05c1) last updated 2016/10/06 10:37:14 (GMT -500)
  lib/ansible/modules/core: (devel b4f6a25195) last updated 2016/10/06 10:37:35 (GMT -500)
  lib/ansible/modules/extras: (useful-error-incorrect-service d6b17be354) last updated 2016/10/06 16:38:39 (GMT -500)
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides

```
##### SUMMARY

Fixes https://github.com/ansible/ansible-modules-extras/issues/2494

```
Before:
$ ansible virt -m firewalld -a "service=httpd permanent=true immediate=true state=enabled" -i /tmp/inventory.txt  -u r
oot
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: dbus.exceptions.DBusException: org.fedoraproject.FirewallD1.Exception: INVALID_SERVICE: 'httpd' not among existing services
192.168.1.100 | FAILED! => {
    "changed": false,
    "failed": true,
    "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_ZRO2lD/ansible_module_firewalld.py\", line 641, in <module>\n    main()\n  File \"/tmp/ansible_ZRO2lD/ansible_module_firewalld.py\", line 418, in main\n    set_service_enabled_permanent(zone, service)\n  File \"/tmp/ansible_ZRO2lD/ansible_module_firewalld.py\", line 278, in set_service_enabled_permanent\n    fw_zone.update(fw_settings)\n  File \"<string>\", line 2, in update\n  File \"/usr/lib/python2.7/site-packages/slip/dbus/polkit.py\", line 121, in _enable_proxy\n    return func(*p, **k)\n  File \"<string>\", line 2, in update\n  File \"/usr/lib/python2.7/site-packages/firewall/client.py\", line 53, in handle_exceptions\n    return func(*args, **kwargs)\n  File \"/usr/lib/python2.7/site-packages/firewall/client.py\", line 441, in update\n    self.fw_zone.update(tuple(settings.settings))\n  File \"/usr/lib/python2.7/site-packages/slip/dbus/proxies.py\", line 51, in __call__\n    return dbus.proxies._ProxyMethod.__call__(self, *args, **kwargs)\n  File \"/usr/lib64/python2.7/site-packages/dbus/proxies.py\", line 145, in __call__\n    **keywords)\n  File \"/usr/lib64/python2.7/site-packages/dbus/connection.py\", line 651, in call_blocking\n    message, timeout)\ndbus.exceptions.DBusException: org.fedoraproject.FirewallD1.Exception: INVALID_SERVICE: 'httpd' not among existing services\n",
    "module_stdout": "",
    "msg": "MODULE FAILURE"
}


After:
$ ansible virt -m firewalld -a "service=httpd permanent=true immediate=true state=enabled" -i /tmp/inventory.txt  -u root
192.168.1.100 | FAILED! => {
    "changed": false,
    "failed": true,
    "msg": "Invalid service name provided, please use services as defined in /etc/services"
}

```

Signed-off-by: Adam Miller maxamillion@fedoraproject.org

### CODE REVIEW

**Comment by bcoca** [comment]
wouldn't a better fix just trap errors from this?

```
try:
     if service in fw.getServices(zone):
```

as you seem to do extra work here to prevent only one of several possible exceptions by these calls.

**Comment by maxamillion** [comment]
@bcoca I don't think that would cause an error to trap, because the service would just simply not be in the active services lists for that zone. I could alternatively do a `fw.getServices()` and get a complete list but doesn't work for permanent operations (for whatever reason there's no equivalent call in the firewalld python API for permanent operations, which will be a hard requirement for offline operations https://github.com/ansible/ansible-modules-extras/pull/3108)

**Comment by bcoca** [comment]
sorry, meant in  `set_service_enabled_permanent`, but they both can give exceptions

**Comment by bcoca** [comment]
probably all the calls should be trapped, this module is extremely optimistic ....

**Comment by gregdek** [comment]
Thanks @maxamillion for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

**Comment by maxamillion** [comment]
I also added a fix for https://github.com/ansible/ansible-modules-extras/issues/236

**Comment by maxamillion** [comment]
@bcoca please review

**Comment by maxamillion** [comment]
After this I have a pretty big refactor that will fix remaining issues other than feature requests but it will make adding new features more striaght-foward.

**Comment by maxamillion** [comment]
shipit

File: system/firewalld.py [hunk 1]
```diff
@@ -104,15 +104,49 @@
...
+    logic and emit (hopefully) useful error messages
+    """
+
+    msgs = []
+
+    try:
+        return action_func(*action_func_args)
+    except Exception:
+        # Make python 2.4 shippable ci tests happy
+        e = sys.exc_info()[1]
```
Comment (abadger) [inline system/firewalld.py:137]: Can use our helper function so people don't have to remember what sys.exc_info[1] means:

```
from ansible.module_utils.pycompat24 import get_exception
[...]
e = get_exception()
```

It does the same thing.

File: system/firewalld.py [hunk 2]
```diff
@@ -104,15 +104,49 @@
...
+        # context for to help the users diagnose what's wrong. Handle that here
+        if "INVALID_SERVICE" in "%s" % e:
+            msgs.append("Services are defined by port/tcp relationship and named as they are in /etc/services (on most systems)")
 
+        if len(msgs) > 0:
+            module.fail_json(
+                msg='ERROR: Exception caught: %s %s' % (e, ', '.join(msgs))
+            )
+        else:
+            module.fail_json(msg='ERROR: Exception caught: %s' % e)
```
Comment (abadger) [inline system/firewalld.py:149]: You may also want to set the exception field in exceptions that are unexpected in the code for easier debugging:

```
import traceback
[...]
module.fail_json([...], exception=traceback.format_exc(e))
```

**Comment by abadger** [comment]
Merged to devel!  Thanks for your patience :-)

**Pull request merged**